File: android/app/src/main/java/com/sendprobe/autolock/storage/AppSettings.kt
package com.sendprobe.autolock.storage

import android.content.Context
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue

/**
 * App-wide settings backed by SharedPreferences, exposed as Compose state
 * so the Settings screen recomposes on change. Centralized here (rather
 * than scattered per-screen state) so the notification listener added in
 * a later step reads the same values the Settings screen writes.
 */
class AppSettings private constructor(context: Context) {
    private val prefs = context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    var dryRunEnabled: Boolean by mutableStateOf(prefs.getBoolean(KEY_DRY_RUN, true))
        private set

    var debounceSeconds: Float by mutableFloatStateOf(prefs.getFloat(KEY_DEBOUNCE, 90f))
        private set

    fun updateDryRunEnabled(value: Boolean) {
        dryRunEnabled = value
        prefs.edit().putBoolean(KEY_DRY_RUN, value).apply()
    }

    fun updateDebounceSeconds(value: Float) {
        debounceSeconds = value
        prefs.edit().putFloat(KEY_DEBOUNCE, value).apply()
    }

    var lastTriggerMillis: Long
        get() = prefs.getLong(KEY_LAST_TRIGGER, 0L)
        set(value) { prefs.edit().putLong(KEY_LAST_TRIGGER, value).apply() }

    /**
     * Returns true if enough time has passed since the last trigger to
     * allow another lock attempt, and records [now] as the new
     * last-trigger time as a side effect when it does.
     */
    fun shouldProceedPastDebounce(now: Long = System.currentTimeMillis()): Boolean {
        val last = lastTriggerMillis
        if (last != 0L && now - last < debounceSeconds * 1000L) {
            return false
        }
        lastTriggerMillis = now
        return true
    }

    companion object {
        private const val PREFS_NAME = "com.sendprobe.autolock.settings"
        private const val KEY_DRY_RUN = "dryRunEnabled"
        private const val KEY_DEBOUNCE = "debounceSeconds"
        private const val KEY_LAST_TRIGGER = "lastTriggerMillis"

        @Volatile private var instance: AppSettings? = null

        fun get(context: Context): AppSettings =
            instance ?: synchronized(this) {
                instance ?: AppSettings(context).also { instance = it }
            }
    }
}

File: ios/Autolock/Storage/AppSettings.swift
import Foundation
import Combine

/// App-wide settings backed by UserDefaults. Centralized (rather than
/// scattered @AppStorage) so the App Intent added in a later step can read
/// the same values the Settings screen writes.
@MainActor
final class AppSettings: ObservableObject {
    static let shared = AppSettings()

    private enum Keys {
        static let dryRunEnabled = "dryRunEnabled"
        static let debounceSeconds = "debounceSeconds"
        static let lastTriggerDate = "lastTriggerDate"
    }

    private let defaults: UserDefaults

    @Published var dryRunEnabled: Bool {
        didSet { defaults.set(dryRunEnabled, forKey: Keys.dryRunEnabled) }
    }

    @Published var debounceSeconds: Double {
        didSet { defaults.set(debounceSeconds, forKey: Keys.debounceSeconds) }
    }

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        // Dry run defaults to true (opt-in to live commands) even on first
        // launch, since UserDefaults.bool(forKey:) already returns false for
        // an unset key -- registerDefaults makes that explicit either way.
        defaults.register(defaults: [
            Keys.dryRunEnabled: true,
            Keys.debounceSeconds: 90.0
        ])
        self.dryRunEnabled = defaults.bool(forKey: Keys.dryRunEnabled)
        self.debounceSeconds = defaults.double(forKey: Keys.debounceSeconds)
    }

    var lastTriggerDate: Date? {
        get { defaults.object(forKey: Keys.lastTriggerDate) as? Date }
        set { defaults.set(newValue, forKey: Keys.lastTriggerDate) }
    }

    /// Returns true if enough time has passed since the last trigger to
    /// allow another lock attempt, and records `now` as the new last-trigger
    /// time as a side effect when it does.
    func shouldProceedPastDebounce(now: Date = Date()) -> Bool {
        if let last = lastTriggerDate, now.timeIntervalSince(last) < debounceSeconds {
            return false
        }
        lastTriggerDate = now
        return true
    }
}
